service/worker/scanner/executions/current_execution.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/uber/cadence/common/reconciliation/invariant"
+	"github.com/uber/cadence/common/reconciliation/store"
+	"github.com/uber/cadence/common/service/dynamicconfig"
+	"github.com/uber/cadence/service/worker/scanner/shardscanner"
+)
+
+const (
+	CurrentExecutionsScannerWFID         = "cadence-sys-current-executions-scanner"
+	CurrentExecutionsScannerWFTypeName   = "cadence-sys-current-executions-scanner-workflow"
+	CurrentExecutionsScannerTaskListName = "cadence-sys-current-executions-scanner-tasklist-0"
+
+	CurrentExecutionsFixerWFTypeName   = "cadence-sys-current-executions-fixer-workflow"
+	CurrentExecutionsFixerTaskListName = "cadence-sys-current-executions-fixer-tasklist-0"
+)
+
+// CurrentScannerWorkflow is the workflow that scans over all current executions
+func CurrentScannerWorkflow(
+	ctx workflow.Context,
+	params shardscanner.ScannerWorkflowParams,
+) error {
+	wf, err := shardscanner.NewScannerWorkflow(ctx, CurrentExecutionsScannerWFTypeName, params)
+	if err != nil {
+		return err
+	}
+
+	return wf.Start(ctx)
+}
+
+// CurrentExecutionsHooks provides hooks for current executions scanner.
+func CurrentExecutionsHooks() *shardscanner.ScannerHooks {
+	wf, err := shardscanner.NewScannerHooks(CurrentExecutionManager, CurrentExecutionIterator)
+	if err != nil {
+		return nil
+	}
+	wf.SetConfig(CurrentExecutionConfig)
+
+	return wf
+}
+
+func CurrentExecutionManager(
+	ctx context.Context,
+	pr persistence.Retryer,
+	params shardscanner.ScanShardActivityParams,
+	_ shardscanner.ScannerConfig,
+) invariant.Manager {
+	var ivs []invariant.Invariant
+	collections := ParseCollections(params.ScannerConfig)
+	for _, fn := range CurrentExecutionType.ToInvariants(collections) {
+		ivs = append(ivs, fn(pr))
+	}
+	return invariant.NewInvariantManager(ivs)
+}
+
+// CurrentFixerWorkflow starts current executions fixer.
+func CurrentFixerWorkflow(
+	ctx workflow.Context,
+	params shardscanner.FixerWorkflowParams,
+) error {
+	wf, err := shardscanner.NewFixerWorkflow(ctx, CurrentExecutionsFixerWFTypeName, params)
+	if err != nil {
+		return err
+	}
+	return wf.Start(ctx)
+}
+
+// CurrentExecutionConfig resolves dynamic config for current executions scanner.
+func CurrentExecutionConfig(ctx shardscanner.Context) shardscanner.CustomScannerConfig {
+	res := shardscanner.CustomScannerConfig{}
+
+	if ctx.Config.DynamicCollection.GetBoolProperty(dynamicconfig.CurrentExecutionsScannerInvariantCollectionHistory, true)() {
+		res[invariant.CollectionHistory.String()] = strconv.FormatBool(true)
+	}
+	if ctx.Config.DynamicCollection.GetBoolProperty(dynamicconfig.CurrentExecutionsScannerInvariantCollectionMutableState, true)() {
+		res[invariant.CollectionMutableState.String()] = strconv.FormatBool(true)
+	}
+
+	return res
+}
+
+// CurrentExecutionFixerHooks provides hooks for current executions fixer.
+func CurrentExecutionFixerHooks() *shardscanner.FixerHooks {
+	h, err := shardscanner.NewFixerHooks(FixerManager, CurrentExecutionFixerIterator)
```

Comment:
reading into this code in depth lately, and... yea, this has never worked.  and still doesn't today.

the scanner reads
```go
	wf, err := shardscanner.NewScannerHooks(CurrentExecutionManager, CurrentExecutionIterator)
```
which scans with the current execution invariants:
```go
	for _, fn := range CurrentExecutionType.ToInvariants(collections) {
		ivs = append(ivs, fn(pr))
	}
```
but the fixer is `FixerManager`, which means fixes work on the _concrete_ invariants:
```go
// FixerManager provides invariant manager for concrete execution fixer.
func FixerManager(_ context.Context, pr persistence.Retryer, _ shardscanner.FixShardActivityParams, _ shardscanner.ScannerConfig) invariant.Manager {
	var ivs []invariant.Invariant
	var collections []invariant.Collection

	collections = append(collections, invariant.CollectionHistory, invariant.CollectionMutableState)

	for _, fn := range ConcreteExecutionType.ToInvariants(collections) {
		ivs = append(ivs, fn(pr))
	}
	return invariant.NewInvariantManager(ivs)
}
```
which causes the fixer to fail predictably:
```
{
  "Input": {
    ...
    "Result": {
      "CheckResultType": "corrupted",
      "DeterminingInvariantType": "concrete_execution_exists",
      "CheckResults": [
        {
          "CheckResultType": "corrupted",
          "InvariantName": "concrete_execution_exists",  <- correct invariant during scan
          "Info": "fake corrupt",                        <- faked for testing
          "InfoDetails": ""
        }
      ]
    }
  },
  "Result": {
    "FixResultType": "failed",
    "DeterminingInvariantName": "history_exists",
    "FixResults": [
      {
        "FixResultType": "failed",
        "InvariantName": "history_exists",
        "CheckResult": {
          "CheckResultType": "failed",
          "InvariantName": "history_exists",                       <- wrong invariant during fix
          "Info": "failed to check: expected concrete execution",  <- wrong entity type
          "InfoDetails": ""
        },
        "Info": "failed fix because check failed",
        "InfoDetails": ""
      },
      {
        "FixResultType": "failed",
        "InvariantName": "open_current_execution",
        "CheckResult": {
          "CheckResultType": "failed",
          "InvariantName": "open_current_execution",               <- wrong invariant during fix
          "Info": "failed to check: expected concrete execution",  <- wrong entity type
          "InfoDetails": ""
        },
        "Info": "failed fix because check failed",
        "InfoDetails": ""
      }
    ]
  }

}
```
It still runs the workflow and activities and creates the correct _number_ of blobstore files (covered in test in description), but the files related to this fixer are `uuid.failed`, not `uuid.fixed`.